0014-http-ilp/0014-http-ilp.md

```diff
@@ -0,0 +1,249 @@
+
+![Sequence Diagram](sequence.png)
+
+### 2.1. Client generates a server-specific token from the hostname and its local secret:
+
+The uploader tool contains an HTTP-ILP client module. This client module has already generated and locally saved a `client_secret` using 256 bits of cryptographically secure randomness.
+
+Before making the paid HTTP request, the HTTP-ILP client module generates a **token** using the `hostname` of the server it is about to make a request to:
+
+* Token: `HMAC(SHA256, client_secret, hostname)`
+
+### 2.1. Client attempts to call API using token
+
+The paid HTTP request is fired off:
+
+``` http
+POST /upload HTTP/1.1
+Host: myservice.example
+Pay-Token: 7y0SfeN7lCuq0GFF5UsMYZofIjJ7LrvPvsePVWSv450
+```
+```
+[...]
+```
+
+Note that the client hasn't paid at this point and is only making the request to solicit a response from the server.
+
+### 2.2. Server rejects request, because token is new and unfunded
+
+The server returns an HTTP error of `402 Payment Required` and includes response headers showing the amount (how much the request would have cost), an ILP address and a condition seed.
+
+``` http
+HTTP/1.1 402 Payment Required
+Pay: 10 us.nexus.ankita.~recv.filepay SkTcFTZCBKgP6A6QOUVcwWCCgYIP4rJPHlIzreavHdU
+Pay-Balance: 0
+```
+
+The client may use the condition seed to create a condition to pay this host. The condition seed is generated by the server as follows:
+
+* Condition Seed: `HMAC(SHA256, receiver_secret, token)`
+
+The condition_seed is now a shared secret between the client and server, but will be unknown to any third-party connectors between them.
+
+That means that if the payment fails, the client can blame the server (assuming it considers itself honest and not compromised.) The client can't prove any wrongdoing to a third party, but since amounts are likely to be very low, this probably wouldn't be worth doing anyway. Also, proving wrongdoing would also involve proving that the server did not provide the agreed-upon service, which is probably near impossible in most cases anyway. However, since it can detect malicious behavior, the client can blacklist malicious hosts and not deal with them in the future.
+
+
+### 2.3. Client initiates an ILP payment to refill its balance
+
+In order to refill its balance, the client now creates an ILP payment with the following properties:
+
+* Destination: `us.nexus.ankita.~recv.filepay`
+* Amount: `100`
+* Condition: `SHA256(fulfillment)`
+* Memo: `pay_token`
+
+The `fulfillment` is generated as:
+
+* Fulfillment: `HMAC(SHA256, condition_seed, destination_account || destination_amount)`
+
+When the prepared payment reaches the receiving wallet, the webhook that is registered with the ankita+filepay@nexus.justmoon.com receiver is triggered. Note that the wallet does not fulfill the payment yet.
+
+### 2.4. Wallet POSTs a payment notification
```

Comment:
Why not just have the server connect to the ILSP using something like BTP, rather than creating yet another encoding for ILP payment details? Websockets might not be ideal but I think it would be better if the server itself were a standard ILP receiver using some ledger protocol to talk to the ILSP.